core/src/com/unciv/logic/automation/NextTurnAutomation.kt

```diff
@@ -309,6 +312,62 @@ class NextTurnAutomation{
         return cityDistances.minBy { it.arialDistance }!!
     }
 
+    private fun offerDeclarationOfFriendship(civInfo: CivilizationInfo) {
+        val canDeclareFriendshipCiv = civInfo.getKnownCivs()
+                .asSequence()
+                .filter { it.isMajorCiv() }
+                .filter { !it.isAtWarWith(civInfo) }
+                .filter { it.getDiplomacyManager(civInfo).relationshipLevel() > RelationshipLevel.Neutral }
+                .filter { !civInfo.getDiplomacyManager(it).hasFlag(DiplomacyFlags.DeclarationOfFriendship) }
+                .filter { !civInfo.getDiplomacyManager(it).hasFlag(DiplomacyFlags.Denunceation) }
+                .sortedByDescending { it.getDiplomacyManager(civInfo).relationshipLevel() }
+                .toList()
+        for (civ in canDeclareFriendshipCiv) {
+            // Default setting is 5, this will be changed according to different civ.
+            if ((1..10).random()<=5) civInfo.getDiplomacyManager(civ).signDeclarationOfFriendship()
+            else continue
+        }
+    }
+
+    private fun offerResearchAgreement(civInfo: CivilizationInfo) {
+        // if Civ has researched future tech, they will not want to sign RA.
+        if (civInfo.tech.getTechUniques().contains("Enables Research agreements")
+                && civInfo.gold >= civInfo.goldCostOfSignResearchAgreement()
+                && civInfo.gameInfo.ruleSet.technologies.values.any { !civInfo.tech.isResearched(it.name) && civInfo.tech.canBeResearched(it.name) }){
+            val canSignResearchAgreementCiv = civInfo.getKnownCivs()
+                    .asSequence()
+                    .filter { it.isMajorCiv() }
+                    .filter { civInfo.getDiplomacyManager(it).hasFlag(DiplomacyFlags.DeclarationOfFriendship) }
```

Comment:
if we implement the canSignResearchAgreement function, we can use it here as well and reduce code duplication